// @scullyio/scully/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.prod = exports.setMyConfig = exports.getMyConfig = exports.launchedBrowser$ = exports.renderRoute = exports.WriteToStorage = exports.routeDiscovery = exports.handleTravesal = exports.loadConfig = exports.staticServer = exports.createFolderFor = exports.findPlugin = exports.getPluginConfig = exports.setPluginConfig = exports.getHandledRoutes = exports.updateScullyConfig = exports.startScully = exports.scullyConfig = exports.RouteTypes = exports.routeSplit = exports.replaceFirstRouteParamWithVal = exports.registerPlugin = exports.httpGetJson = exports.configValidator = void 0;
const pluginConfig_1 = require("./lib/pluginManagement/pluginConfig");
Object.defineProperty(exports, "findPlugin", { enumerable: true, get: function () { return pluginConfig_1.findPlugin; } });
Object.defineProperty(exports, "getMyConfig", { enumerable: true, get: function () { return pluginConfig_1.getConfig; } });
Object.defineProperty(exports, "getPluginConfig", { enumerable: true, get: function () { return pluginConfig_1.getPluginConfig; } });
Object.defineProperty(exports, "setMyConfig", { enumerable: true, get: function () { return pluginConfig_1.setConfig; } });
Object.defineProperty(exports, "setPluginConfig", { enumerable: true, get: function () { return pluginConfig_1.setPluginConfig; } });
const pluginRepository_1 = require("./lib/pluginManagement/pluginRepository");
Object.defineProperty(exports, "configValidator", { enumerable: true, get: function () { return pluginRepository_1.configValidator; } });
Object.defineProperty(exports, "registerPlugin", { enumerable: true, get: function () { return pluginRepository_1.registerPlugin; } });
require("./lib/pluginManagement/systemPlugins");
const executePlugins_1 = require("./lib/renderPlugins/executePlugins");
Object.defineProperty(exports, "renderRoute", { enumerable: true, get: function () { return executePlugins_1.renderRoute; } });
const launchedBrowser_1 = require("./lib/renderPlugins/launchedBrowser");
Object.defineProperty(exports, "launchedBrowser$", { enumerable: true, get: function () { return launchedBrowser_1.launchedBrowser$; } });
const writeToFs_plugin_1 = require("./lib/systemPlugins/writeToFs.plugin");
Object.defineProperty(exports, "WriteToStorage", { enumerable: true, get: function () { return writeToFs_plugin_1.WriteToStorage; } });
const utils_1 = require("./lib/utils");
Object.defineProperty(exports, "createFolderFor", { enumerable: true, get: function () { return utils_1.createFolderFor; } });
const cli_options_1 = require("./lib/utils/cli-options");
Object.defineProperty(exports, "prod", { enumerable: true, get: function () { return cli_options_1.prod; } });
const config_1 = require("./lib/utils/config");
Object.defineProperty(exports, "loadConfig", { enumerable: true, get: function () { return config_1.loadConfig; } });
Object.defineProperty(exports, "scullyConfig", { enumerable: true, get: function () { return config_1.scullyConfig; } });
Object.defineProperty(exports, "updateScullyConfig", { enumerable: true, get: function () { return config_1.updateScullyConfig; } });
require("./lib/utils/exitHandler");
const handleTravesal_1 = require("./lib/utils/handlers/handleTravesal");
Object.defineProperty(exports, "handleTravesal", { enumerable: true, get: function () { return handleTravesal_1.handleTravesal; } });
const routeDiscovery_1 = require("./lib/utils/handlers/routeDiscovery");
Object.defineProperty(exports, "routeDiscovery", { enumerable: true, get: function () { return routeDiscovery_1.routeDiscovery; } });
const httpGetJson_1 = require("./lib/utils/httpGetJson");
Object.defineProperty(exports, "httpGetJson", { enumerable: true, get: function () { return httpGetJson_1.httpGetJson; } });
const interfacesandenums_1 = require("./lib/utils/interfacesandenums");
Object.defineProperty(exports, "RouteTypes", { enumerable: true, get: function () { return interfacesandenums_1.RouteTypes; } });
const replaceFirstRouteParamWithVal_1 = require("./lib/utils/replaceFirstRouteParamWithVal");
Object.defineProperty(exports, "replaceFirstRouteParamWithVal", { enumerable: true, get: function () { return replaceFirstRouteParamWithVal_1.replaceFirstRouteParamWithVal; } });
const routeSplit_1 = require("./lib/utils/routeSplit");
Object.defineProperty(exports, "routeSplit", { enumerable: true, get: function () { return routeSplit_1.routeSplit; } });
const staticServer_1 = require("./lib/utils/serverstuff/staticServer");
Object.defineProperty(exports, "staticServer", { enumerable: true, get: function () { return staticServer_1.staticServer; } });
const routeStorage_1 = require("./lib/utils/services/routeStorage");
Object.defineProperty(exports, "getHandledRoutes", { enumerable: true, get: function () { return routeStorage_1.getHandledRoutes; } });
const startup_1 = require("./lib/utils/startup");
Object.defineProperty(exports, "startScully", { enumerable: true, get: function () { return startup_1.startScully; } });
__exportStar(require("./lib/utils/log"), exports);
