// @scullyio/scully/src/lib/utils/cli-options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.options = exports.killServer = exports.serve = exports.disableProjectFolderCheck = exports.stats = exports.watch = exports.tds = exports.sslKey = exports.sslCert = exports.ssl = exports.showBrowser = exports.sge = exports.serverTimeout = exports.scanRoutes = exports.routeFilter = exports.removeStaticDist = exports.proxyConfigFile = exports.project = exports.prod = exports.port = exports.pluginFolder = exports.pluginsError = exports.pjFirst = exports.path = exports.openNavigator = exports.noPrompt = exports.noLog = exports.logSeverity = exports.hostName = exports.handle404 = exports.folder = exports.configFileName = exports.baseFilter = void 0;
const yargs_1 = __importDefault(require("yargs"));
_a = 
/** return the argv */
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
yargs_1.default
    /** watch mode */
    .boolean('w')
    .default('w', false)
    .alias('w', 'watch')
    .describe('w', 'Use this flag for use the watch mode into scully')
    /** remove Static dist */
    .boolean('RSD')
    .default('RSD', false)
    .alias('RSD', 'removeStaticDist')
    .describe('RSD', 'Use this flag to remove the Scully outfolder before starting')
    /** open browser */
    .boolean('o')
    .default('o', false)
    .alias('o', 'openNavigator')
    .alias('open', 'openNavigator')
    .describe('o', 'Use this flag for open the browser with the serve')
    /** ssl */
    .boolean('ssl')
    .default('ssl', false)
    .describe('ssl', 'Add self ssl into the server')
    .string('sslKey')
    .default('sslKey', undefined)
    .alias('ssl-key', 'sslKey')
    .describe('sslKey', 'Add ssl-key into the server')
    .string('sslCert')
    .default('sslCert', undefined)
    .alias('ssl-cert', 'sslCert')
    .describe('sslCert', 'Add ssl-cert into the server')
    /** test data server */
    .boolean('tds')
    .default('tds', false)
    .describe('tds', 'start TestDataServer')
    /** proxy config */
    .string('proxy')
    .default('proxy', undefined)
    .alias('proxy', 'proxyConfigFile')
    .alias('proxy', 'proxyConfig')
    .alias('proxy-config', 'proxyConfig')
    .describe('proxy', 'Load proxy config from file')
    /** hostName */
    .string('hostName')
    .default('hostName', undefined)
    .alias('host', 'hostName')
    .describe('host', 'Add hostname for scully')
    /** path */
    .string('path')
    .default('path', undefined)
    .describe('path', 'The path to generate')
    /** port */
    .number('port')
    .default('port', undefined)
    .alias('p', 'port')
    .describe('port', 'The port to run on')
    /** showBrowser */
    .boolean('sb')
    .alias('sb', 'showBrowser')
    .describe('sb', 'Shows the puppeteer controlled browser')
    /** folder */
    .string('folder')
    .default('folder', undefined)
    .describe('folder', 'folder')
    /** showGuessErrors */
    .boolean('sge')
    .alias('sge', 'showGuessError')
    .describe('sge', 'dumps the error from guess to the console')
    .string('cf')
    .alias('cf', 'configFile')
    .alias('cf', 'configFileName')
    .default('cf', '')
    .describe('cf', 'provide name of the config file to use. if the option --project is also there that takes precedence)')
    /** don't log info lines to console */
    .boolean('nl')
    .default('nl', false)
    .alias('nl', 'noLog')
    .describe('nl', 'Log warnings and errors only')
    /** projectName */
    .string('pr')
    .alias('pr', 'project')
    .default('pr', '')
    .describe('pr', 'provide name of the project to handle')
    /** scanRoutes */
    .boolean('sr')
    .default('sr', false)
    .alias('sr', 'scanRoutes')
    .alias('sr', 'scan')
    .describe('sr', 'Scan the app for unhandled routes')
    /** baseFilter */
    .string('bf')
    .alias('bf', 'baseFilter')
    .default('bf', '')
    .describe('bf', 'provide a wildcard string separated by ,(comma) to filter the unhandled routes')
    /** filter */
    .string('routeFilter')
    .alias('routeFilter', 'rf')
    .alias('routeFilter', 'route-filter')
    .default('routeFilter', '')
    .describe('routeFilter', 'provide a wildcard string separated by ,(comma) to filter the handled routes')
    /** server Timout */
    .number('st')
    .default('st', 0)
    .alias('st', 'serverTimeout')
    .describe('st', 'The time Scully will wait for the server before timeout. in milliseconds')
    /** package json fist */
    .boolean('pjf')
    .default('pjf', false)
    .alias('pjf', 'pjFirst')
    .alias('pjf', 'pj-first')
    .describe('pjf', 'Scan for package.json first instead of angular.json')
    /** handle 404 in server */
    .choices('handle404', ['', 'index', 'baseOnly', '404', 'none'])
    .alias('handle404', '404')
    .default('handle404', '')
    .describe('handle404', 'determine how 404 is handled')
    /** pluginFolder, the place to look for custom config/plugins */
    .string('pf')
    .alias('pf', 'pluginFolder')
    .alias('pf', 'plugin-folder')
    .default('pf', './scully')
    .describe('pf', 'Folder to look for custom config/plugins (will use the ts-config in there to compile)')
    /** Exit Scully with plugin error */
    .boolean('pe')
    .alias('pe', 'pluginsError')
    .default('pe', true)
    .describe('pe', "Exit scully's run when exist an error in a plugin")
    /** No use prompts (for use in undetectable CI/CD) */
    .boolean('np')
    .alias('np', 'noPrompt')
    .alias('np', 'no-prompt')
    .default('np', false)
    .describe('np', 'Do not pause for user input')
    /** Use Prod Mode */
    .boolean('prod')
    .alias('prod', 'Production')
    .default('prod', false)
    .describe('prod', 'Use prod mode for Scully')
    /** logSeverity */
    .choices('logSeverity', ['normal', 'warning', 'error', 'none'])
    .alias('logSeverity', 'ls')
    .alias('logSeverity', 'log-severity')
    .default('logSeverity', 'warning')
    .describe('logSeverity', 'select the log-severity level')
    /** write Statistics to scullyStats.json */
    .boolean('stats')
    .alias('stats', 'statistics')
    .default('stats', false)
    .describe('stats', 'Output select statistics to sucllyStats.json')
    /** Don't check if Scully was launched from project folder */
    .boolean('disableProjectFolderCheck')
    .default('disableProjectFolderCheck', false)
    .describe('disableProjectFolderCheck', "Don't check if Scully was launched from project folder").argv, 
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
exports.baseFilter = _a.baseFilter, exports.configFileName = _a.configFileName, exports.folder = _a.folder, exports.handle404 = _a.handle404, exports.hostName = _a.hostName, exports.logSeverity = _a.logSeverity, exports.noLog = _a.noLog, exports.noPrompt = _a.noPrompt, exports.openNavigator = _a.openNavigator, exports.path = _a.path, exports.pjFirst = _a.pjFirst, exports.pluginsError = _a.pluginsError, exports.pluginFolder = _a.pluginFolder, exports.port = _a.port, exports.prod = _a.prod, exports.project = _a.project, exports.proxyConfigFile = _a.proxyConfigFile, exports.removeStaticDist = _a.removeStaticDist, exports.routeFilter = _a.routeFilter, exports.scanRoutes = _a.scanRoutes, exports.serverTimeout = _a.serverTimeout, exports.sge = _a.sge, exports.showBrowser = _a.showBrowser, exports.ssl = _a.ssl, exports.sslCert = _a.sslCert, exports.sslKey = _a.sslKey, exports.tds = _a.tds, exports.watch = _a.watch, exports.stats = _a.stats, exports.disableProjectFolderCheck = _a.disableProjectFolderCheck;
yargs_1.default.help();
const commandsArray = yargs_1.default.argv._.filter((c) => typeof c === 'string').map((c) => c.toLowerCase().trim());
exports.serve = commandsArray.includes('serve');
exports.killServer = commandsArray.includes('killserver');
exports.options = yargs_1.default.option('port', {
    alias: 'p',
    type: 'number',
    description: 'The port to run on',
}).argv;
