// @scullyio/ng-lib/esm2015/lib/config/scully-config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/config/scully-config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/**
 * @record
 */
export function ScullyLibConfig() { }
if (false) {
    /** @type {?|undefined} */
    ScullyLibConfig.prototype.useTransferState;
    /** @type {?|undefined} */
    ScullyLibConfig.prototype.alwaysMonitor;
    /** @type {?|undefined} */
    ScullyLibConfig.prototype.manualIdle;
    /** @type {?|undefined} */
    ScullyLibConfig.prototype.baseURIForScullyContent;
}
/** @type {?} */
export const ScullyDefaultSettings = {
    useTransferState: true,
    alwaysMonitor: false,
    manualIdle: false,
    baseURIForScullyContent: 'http://localhost:1668',
};
/** @type {?} */
export const SCULLY_LIB_CONFIG = new InjectionToken('scullyLibConfig', {
    factory: (/**
     * @return {?}
     */
    () => ScullyDefaultSettings),
});
